src/enclave/enclave_oe.c

```diff
@@ -1,27 +1,144 @@
+    const sgxlkl_enclave_config_t* cfg = state->config;
+
+    // import host envp
+    state->num_imported_env = 0;
+    state->imported_env = NULL;
+
+    if (sgxlkl_enclave_state.shared_memory.env && cfg->num_host_import_env > 0)
+    {
+        state->imported_env = malloc(sizeof(char*) * cfg->num_host_import_env);
+        if (!state->imported_env)
+            sgxlkl_fail("out of memory\n");
+
+        for (size_t i = 0; i < cfg->num_host_import_env; i++)
+        {
+            const char* name = cfg->host_import_env[i];
+            for (char* const* p = sgxlkl_enclave_state.shared_memory.env;
+                 p && *p != NULL;
+                 p++)
+            {
+                size_t n = strlen(name);
+                if (strncmp(name, *p, n) == 0 && (*p)[n] == '=')
+                {
+                    const char* str = *p;
+                    size_t len = strlen(str);
+                    char* cpy = malloc(len + 1);
+                    if (!cpy)
+                        sgxlkl_fail("out of memory\n");
+                    memcpy(cpy, str, len + 1);
+                    state->imported_env[state->num_imported_env++] = cpy;
+                }
+            }
+        }
+    }
+
+    size_t total_size = 0;
+    size_t total_count = 1;
+    for (size_t i = 0; i < cfg->num_args; i++)
+        total_size += strlen(cfg->args[i]) + 1;
+    total_count += cfg->num_args + 1;
+    for (size_t i = 0; i < cfg->num_env; i++)
+        total_size += strlen(cfg->env[i]) + 1;
+    total_count += cfg->num_env + 1;
+    for (size_t i = 0; i < state->num_imported_env; i++)
+        total_size += strlen(state->imported_env[i]) + 1;
+    total_count += state->num_imported_env + 1;
+    total_count += 1; // auxv terminator
+    total_count += 1; // platform-independent stuff terminator
+
+    char* buf = calloc(total_size, sizeof(char));
+    char** out = calloc(total_count, sizeof(char*));
+
+    size_t j = 0;
+    char* buf_ptr = buf;
+
+#define ADD_STRING(S)               \
+    {                               \
+        size_t len = strlen(S) + 1; \
+        memcpy(buf_ptr, (S), len);  \
+        out[j++] = buf_ptr;         \
+        buf_ptr += len;             \
+    }
+
+    elf64_stack_t* stack = &sgxlkl_enclave_state.elf64_stack;
 
-extern _Atomic(enum sgxlkl_libc_state) __libc_state;
+    // argv
+    stack->argv = out;
+    for (size_t i = 0; i < cfg->num_args; i++)
+        ADD_STRING(cfg->args[i]);
+    stack->argc = j;
+    out[j++] = NULL;
 
-sgxlkl_config_t* sgxlkl_enclave;
+    // envp
+    stack->envp = out + j;
+    for (size_t i = 0; i < cfg->num_env; i++)
+        ADD_STRING(cfg->env[i]);
+    for (size_t i = 0; i < state->num_imported_env; i++)
+        // Is this the right order for imported vars?
+        ADD_STRING(state->imported_env[i]);
+    out[j++] = NULL;
+
+    // auxv
+    stack->auxv = (Elf64_auxv_t**)(out + j);
+    for (size_t i = 0; i < cfg->num_auxv; i++)
```

Comment:
Yeah, I've been meaning to ask you about that. Wouldn't the ELF loader fill in all these bits, so that neither the host nor the config need to be aware of auxv?

Reply:
We don't currently use the Linux ELF loader, but that is indeed what should be setting this.  We need to clean this up in the userspace process creation code.

Reply:
Great, I'll remove that setting then!